src/core/lib/iomgr/tcp_client_posix.cc

```diff
@@ -342,13 +410,61 @@ static int64_t tcp_connect(grpc_closure* closure, grpc_endpoint** ep,
     grpc_core::ExecCtx::Run(DEBUG_LOCATION, closure, error);
     return 0;
   }
-  grpc_tcp_client_create_from_prepared_fd(interested_parties, closure, fd,
-                                          channel_args, &mapped_addr, deadline,
-                                          ep);
-  return 0;
+  return grpc_tcp_client_create_from_prepared_fd(interested_parties, closure,
+                                                 fd, channel_args, &mapped_addr,
+                                                 deadline, ep);
 }
 
-static bool tcp_cancel_connect(int64_t /*connection_handle*/) { return false; }
+static bool tcp_cancel_connect(int64_t connection_handle) {
+  if (connection_handle <= 0) {
+    return false;
+  }
+  int shard_number = connection_handle % g_pending_connection_shards;
+  async_connect* ac = nullptr;
+  {
+    grpc_core::MutexLock lock(g_shard_mu[shard_number]);
+    auto it = g_pending_connections[shard_number]->find(connection_handle);
+    if (it != g_pending_connections[shard_number]->end()) {
+      ac = it->second;
+      GPR_ASSERT(ac != nullptr);
+      // Trying to acquire ac->mu here would could cause a deadlock because
+      // the on_writable method tries to acquire the two mutexes used
+      // here in the reverse order. But we dont need to acquire ac->mu before
+      // incrementing ac->refs here. This is because the on_writable
+      // method decrements ac->refs only after deleting the connection handle
+      // from the corresponding hashmap. If the code enters here, it means that
+      // deletion hasn't happened yet. The deletion can only happen after the
+      // corresponding g_shard_mu is unlocked.
+      ++ac->refs;
+      // Remove connection from list of active connections.
+      g_pending_connections[shard_number]->erase(it);
+    }
+  }
+  if (ac == nullptr) {
+    return false;
+  }
+  gpr_mu_lock(&ac->mu);
+  bool connection_cancel_success = (ac->fd != nullptr);
+  if (connection_cancel_success) {
+    // Connection is still pending. The on_writable callback hasn't executed
+    // yet because ac->fd != nullptr.
+    ac->connect_cancelled = true;
+    // Force on_writable to run as soon as possible and gracefully terminate.
+    // When the on_writable callback runs, it would detect that a connect
+    // cancellation is requested and take the necessary actions.
+    grpc_fd_set_writable(ac->fd);
```

Comment:
Could we trigger this by simply shutting down the fd?

Reply:
Yes I think that should also work. I updated it. But the thread calling grpc_tcp_cancel_connect would need to have an ExecCtx now. Is that okay ? Otherwise we can revert back to grpc_fd_set_writable